### PR TITLE
CI: add the sanitizers pipelines (e.g. ASAN) to Buildkite

### PR DESCRIPTION
**Pull request by tkf**
This is a continuation of #41522 and simply "codifies" what is already documented.

@staticfloat @DilumAluthge I mostly just copied llvmpasses. Let me know if I need to fix something.

### CODE REVIEW

**Comment by DilumAluthge** [comment]
Instead of trying to write stuff into `/tmp`, can you just do `make` in-place, like we do in the [llvm passes](https://github.com/JuliaLang/julia/blob/f8d04b27a7514ecd6e9deb4f605f96984075c773/.buildkite/llvm_passes.yml#L47)?

**Comment by tkf** [comment]
Hmm... I kinda want to keep the out-of-tree -based script so that it is easy to try it out. It's also nice to have a separated toolchain directory so that you don't need to re-install the toolchain (e.g., we can have `contrib/tsan.sh` that shares the toolchain with `contrib/asan.sh`).

Maybe we can use `./tmp` (which is already listed in `.gitignore`)?

**Comment by DilumAluthge** [comment]
> Maybe we can use `./tmp` (which is already listed in `.gitignore`)?

Sure, let's give that a try.

**Comment by tkf** [comment]
image

Too long? (Interesting that it's just two seconds less than the time limit)

File: .buildkite/sanitizers.yml
Comment (tkf) [inline .buildkite/sanitizers.yml:15]: Is `julia` command unavailable in `$PATH`? I get:

> contrib/check-asan.jl: line 12: exec: julia: not found

https://buildkite.com/julialang/julia-master/builds/1815#2ea863e5-e927-484d-a64d-0c0bf51346d7

Comment (staticfloat) [reply .buildkite/sanitizers.yml:15]: There are multiple layers of sand boxing happening here. We need to add “/cache” as a mapping to the workspaces to the sandbox plugin. I’m AFK right now otherwise I would do it.

File: .buildkite/sanitizers.yml
Comment (tkf) [inline .buildkite/sanitizers.yml:27]:
```suggestion
      contrib/check-asan.jl ./tmp/test-asan/usr/bin/julia-debug
    workspaces:
      - "/cache/repos:/cache/repos"
```

Like this?

**Comment by DilumAluthge** [comment]
> image
> 
> Too long? (Interesting that it's just two seconds less than the time limit)

We should probably increase the timeout for that specific pipeline to be 120 minutes.

File: .buildkite/sanitizers.yml
Comment (DilumAluthge) [reply .buildkite/sanitizers.yml:15]: Maybe I'm misunderstanding, but which Julia binary are you trying to run here? Presumably you're trying to run the Julia binary that you just built from source, right? So shouldn't you do e.g. `./tmp/julia`?

Comment (tkf) [reply .buildkite/sanitizers.yml:15]: Ah, sorry it was unclear. I want to use `julia` from `JuliaCI/julia#v1` to run the script `check-asan.jl` which then in turn runs `./tmp/test-asan/usr/bin/julia-debug`. I want some stable `julia` binary just for running the script. (I could do it in bash but it's rather too complicated)

**Comment by DilumAluthge** [comment]
It would be good to edit those `echo` commands to be a little more descriptive. For example, what does `contrib/asan.sh` do, and what does `contrib/check-asan.jl` do?

**Review by DilumAluthge — APPROVED** [review]

**Comment by DilumAluthge** [comment]
LGTM.

It would be good to get Elliot's approval before merging.

**Comment by vchuravy** [comment]
One issue here is that the LLVM used is not built with ASAN itself, normally we recommend `USE_BINARYBUILDER_LLVM=0` for the second stage to build LLVM itself with ASAN, but that might be to expensive to do on CI and we would need to maintain a binary cache (like for the assert build)

**Comment by tkf** [comment]
I think it's still nice to make sure our codebase itself is ASAN compatible; i.e., the code path with `JL_ASAN_ENABLED` is compilable and there is no apparent problem in the run-time that can be detected with ASAN.

File: contrib/asan.sh
Comment (staticfloat) [inline contrib/asan.sh:79]: I would prefer that we have a `.buildkite/asan/Make.user.asan` or something that has these values, and then we just `cat .buildkite/asan/Make.user.asan >> Make.user` in the script, to make this a little clearer.

File: .buildkite/sanitizers.yml
Comment (staticfloat) [inline .buildkite/sanitizers.yml:30]: Can we put these things in `.buildkite/asan/` or something like that?  I feel the `contrib` directory is kind of the wild west, and I'd like to make it clearer that these things are used in CI, and that `asan.sh`, `check-asan.jl` and (potentially) `Make.user.asan` all "go together'.

File: contrib/asan.sh
Comment (staticfloat) [inline contrib/asan.sh:15]:
```suggestion
# `$WORKSPACE` is a directory in which we create `toolchain` and `asan`
```

File: contrib/asan.sh
Comment (staticfloat) [inline contrib/asan.sh:36]: We should not assume that we're sitting in the top-level repository directory; we should have something like the following at the top:

```
JULIA_HOME="$(dirname "$( cd "$( dirname "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )" )"
```

And then we should `make -C "${JULIA_HOME}" configure O="${TOOLCHAIN}"` here.

File: contrib/asan.sh
Comment (staticfloat) [inline contrib/asan.sh:40]: I actually didn't realize we could use BB LLVM and get Clang as well.  That is super, super cool.

File: contrib/check-asan.jl
Comment (staticfloat) [inline contrib/check-asan.jl:43]: Love seeing people using `addenv()` :)

**Review by staticfloat — APPROVED** [review]

File: .buildkite/sanitizers.yml
Comment (tkf) [reply .buildkite/sanitizers.yml:30]: I was thinking `asan.sh` and `check-asan.jl` as scripts that some devs may invoke on their local checkout (rather than purely CI scripts). It'd be handy to have "just execute" script when you are first trying out how to use ASAN. It's also very useful for comparing your local setup against what's actually CI'ed. It ultimately doesn't matter if it's in `.buildkite` directory for such usecase but I think `contrib/` may be the "least bad" option for such use case. It may also be slightly more discoverable as it's easy to suspect utility scripts may be in there by the name `contrib/`.

What about moving them to `contrib/asan/build.sh` and  `contrib/asan/check.jl`? We can also create `contrib/asan/Make.user.tools` and `contrib/asan/Make.user.asan`.

Comment (staticfloat) [reply .buildkite/sanitizers.yml:30]: Yep, totally happy with that.

File: contrib/asan/Make.user.asan
Comment (staticfloat) [inline contrib/asan/Make.user.asan:23]: Why do you need to set `JULIA_PRECOMPILE=1`?  Isn't that already the default?

**Review by staticfloat — APPROVED** [review]

File: contrib/asan/Make.user.asan
Comment (tkf) [reply contrib/asan/Make.user.asan]: Thanks, I forgot to remove this.

**Comment by tkf** [comment]
Maybe better not backporting this?
https://github.com/JuliaLang/julia/pull/41554#issuecomment-882711000
https://github.com/JuliaLang/julia/pull/41499#issuecomment-882717880

**Comment by DilumAluthge** [comment]
This is my thinking: the CI PRs that we've been making tend to build on one another. So it will probably be cleanest just to backport all CI PRs.

Once I merge https://github.com/JuliaLang/julia/pull/41606, it can be backported. One of the things that https://github.com/JuliaLang/julia/pull/41606 does is that it only runs `asan` on `master`. So if we backport https://github.com/JuliaLang/julia/pull/41606 to 1.6, then `asan` won't run on the `release-1.6` branch, which I think is the correct behavior.

**Comment by tkf** [comment]
Thanks for the explanation. I saw your comments in https://github.com/JuliaLang/julia/pull/41554#issuecomment-883239841 too and I see that it makes sense either way.

**Pull request merged**
